### PR TITLE
Nutanix cloud provider support

### LINKED ISSUE #8438: Implement Nutanix as KKP cloud provider
**User Story**
Nutanix is a provider for on premise infrastructure that we want KKP to integrate with. To do that, we need to create a new cloud provider implementation in KKP (https://github.com/kubermatic/kubermatic/tree/master/pkg/provider/cloud) that creates cluster-specific resources (e.g. a subnet or a network security group).

Creating nodes should not be supported as part of this ticket, as that will be implemented in `machine-controller`.

What needs to be created:

- Category for KKP clusters and a value for the specific cluster

**Acceptance criteria**
Nutanix is supported as a cloud provider in `Cluster` resources and can be used to provision cloud resources.

### PR DESCRIPTION
**Pull request by embik**
**What this PR does / why we need it**:
Adds support for Nutanix as a cloud provider, including all the necessary scaffolding for a new cloud provider in the API.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #8438

**Special notes for your reviewer**:

- This PR is big enough as it is, so it does not implement necessary KKP API endpoints to query cloud resources for the UI. That's supposed to be added as https://github.com/kubermatic/kubermatic/issues/8679.

**Documentation**:

**Does this PR introduce a user-facing change?**:

```release-note
Add experimental/alpha cloud provider support for Nutanix
```

### CODE REVIEW

File: .golangci.yml [hunk 1]
```diff
@@ -60,6 +60,7 @@ issues:
   - cyclomatic complexity [0-9]+ of func `initTestEndpoint` is high
   - cyclomatic complexity [0-9]+ of func `\(\*Provider\)\.InitializeCloudProvider` is high
   - cyclomatic complexity [0-9]+ of func `\(\*Reconciler\)\.ensureResourcesAreDeployed` is high
+  - cyclomatic complexity [0-9]+ of func `GetAPIV2NodeCloudSpec` is high
```
Comment (xrstf) [inline .golangci.yml:63]: a cute kitten just died

File: .prow.yaml [hunk 2]
```diff
@@ -1093,6 +1093,93 @@ presubmits:
...
+        # docker-in-docker needs privileged mode
+        securityContext:
+          privileged: true
+        resources:
+          requests:
+            memory: 4Gi
+            cpu: 3.5
+          limits:
+            memory: 4Gi
+  - name: pre-kubermatic-e2e-nutanix-centos-1.22
```
Comment (xrstf) [inline .prow.yaml:1140]: please keep newlines between jobs

File: docs/zz_generated.seed.yaml [hunk 3]
```diff
@@ -129,6 +131,23 @@ spec:
...
+          # Images to use for each supported operating system
+          images:
+            amzn2: ""
+            centos: ""
+            flatcar: ""
+            rhel: ""
+            sles: ""
+            ubuntu: ""
+          # Optional: Port to use when connecting to the Nutanix Prism Central endpoint (defaults to 9440)
+          port: null
```
Comment (xrstf) [inline docs/zz_generated.seed.yaml]: Can we show the default value here?

File: pkg/handler/v1/provider/nutanix.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+*/
+
+package provider
+
+import (
+	"context"
+	"net/http"
+	"strconv"
+	// "github.com/go-kit/kit/endpoint"
+	// "k8c.io/kubermatic/v2/pkg/provider"
```
Comment (xrstf) [inline pkg/handler/v1/provider/nutanix.go]: leftovers?

File: pkg/handler/v1/provider/nutanix.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	req.NutanixCommonReq = commonReq.(NutanixCommonReq)
+
+	return req, nil
+}
+
+/*
+func NutanixClusterEndpoint(presetProvider provider.PresetProvider, seedsGetter provider.SeedsGetter, userInfoGetter provider.UserInfoGetter) endpoint.Endpoint {
+
+}
+*/
```
Comment (xrstf) [inline pkg/handler/v1/provider/nutanix.go]: What's this?

File: pkg/provider/cloud/nutanix/client.go [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+		}
+	}
+
+	port := 9440
+	if dc.Port != nil {
+		port = int(*dc.Port)
+	}
+
+	creds := nutanixclient.Credentials{
+		URL:      fmt.Sprintf("%s:%d", dc.Endpoint, port),
```
Comment (xrstf) [inline pkg/provider/cloud/nutanix/client.go]: `net.JoinHostPort` maybe?

File: pkg/provider/cloud/nutanix/client.go [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+	if cloud.ProxyURL == "" {
+		if cloud.CredentialsReference != nil {
+			credsProxyURL, _ := secretKeyGetter(cloud.CredentialsReference, resources.NutanixProxyURL)
+			if credsProxyURL != "" {
+				proxyURL = credsProxyURL
+			}
+		}
+	} else {
+		proxyURL = cloud.ProxyURL
+	}
```
Comment (xrstf) [inline pkg/provider/cloud/nutanix/client.go]: Couldn't this be

```go
proxyURL := cloud.ProxyURL
if proxyURL == "" && cloud.CredentialsReference != nil {
   credsProxyURL, _ := secretKeyGetter(cloud.CredentialsReference, resources.NutanixProxyURL)
   if credsProxyURL != "" {
      proxyURL = credsProxyURL
   }
}
```

File: pkg/provider/cloud/nutanix/provider.go [hunk 8]
```diff
@@ -0,0 +1,292 @@
+/*
+Copyright 2021 The Kubermatic Kubernetes Platform contributors.
```
Comment (xrstf) [inline pkg/provider/cloud/nutanix/provider.go:2]: 2022

File: pkg/provider/cloud/nutanix/provider.go [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+
+func (n *Nutanix) CleanUpCloudProvider(cluster *kubermaticv1.Cluster, update provider.ClusterUpdater) (*kubermaticv1.Cluster, error) {
+	logger := n.log.With("cluster", cluster.Name)
+
+	client, err := GetClientSet(n.dc, cluster.Spec.Cloud.Nutanix, n.secretKeySelector)
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Infow("removing category values")
```
Comment (xrstf) [inline pkg/provider/cloud/nutanix/provider.go]: no `w` needed

File: pkg/provider/cloud/nutanix/provider.go [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+
+func (n *Nutanix) reconcileCluster(cluster *kubermaticv1.Cluster, update provider.ClusterUpdater, force bool) (*kubermaticv1.Cluster, error) {
+	logger := n.log.With("cluster", cluster.Name)
+
+	client, err := GetClientSet(n.dc, cluster.Spec.Cloud.Nutanix, n.secretKeySelector)
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Infow("reconciling category and value")
```
Comment (xrstf) [inline pkg/provider/cloud/nutanix/provider.go]: no `w` needed

File: pkg/provider/kubernetes/credentials.go [hunk 11]
```diff
@@ -476,3 +479,37 @@ func createOrUpdateAnexiaSecret(ctx context.Context, seedClient ctrlruntimeclien
...
+		resources.NutanixUsername: []byte(spec.Username),
+		resources.NutanixPassword: []byte(spec.Password),
+	}
+
+	if spec.ProxyURL != "" {
+		secretData[resources.NutanixProxyURL] = []byte(spec.ProxyURL)
+	}
+
+	credentialRef, err := ensureCredentialSecret(ctx, seedClient, cluster, secretData)
+
```
Comment (xrstf) [inline pkg/provider/kubernetes/credentials.go]: this empty line can go

**Review by xrstf — CHANGES_REQUESTED** [review]
consider yourself nutpicked

File: .golangci.yml (hunk 1 above)
Comment (embik) [reply .golangci.yml:63]: Even that cute kitten can understand this function as it's dead simple, so I honestly don't feel too bad.

File: docs/zz_generated.seed.yaml (hunk 3 above)
Comment (embik) [reply docs/zz_generated.seed.yaml]: Done!

File: pkg/handler/v1/provider/nutanix.go (hunk 5 above)
Comment (embik) [reply pkg/handler/v1/provider/nutanix.go]: Leftover. It's gone now.

File: pkg/provider/cloud/nutanix/client.go (hunk 6 above)
Comment (embik) [reply pkg/provider/cloud/nutanix/client.go]: Sure!

File: pkg/handler/v1/provider/nutanix.go (hunk 4 above)
Comment (embik) [reply pkg/handler/v1/provider/nutanix.go]: Yes, removed.

File: pkg/provider/cloud/nutanix/client.go (hunk 7 above)
Comment (embik) [reply pkg/provider/cloud/nutanix/client.go]: Yeah you're right, that's better. Thanks!

**Review by xrstf — APPROVED** [review]
/approve
/hold

pls unhold when docker rate limitting is over

**Pull request merged**